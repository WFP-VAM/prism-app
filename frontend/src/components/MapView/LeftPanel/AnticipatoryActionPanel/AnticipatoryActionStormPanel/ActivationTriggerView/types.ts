import { AACategory } from 'context/anticipatoryAction/AAStormStateSlice/parsedStormDataTypes';

export const AAPanelCategories: AACategory[] = [
  AACategory.Severe,
  AACategory.Moderate,
];

export const AADisplayCategory: {
  [key in AACategory]?: string;
} = {
  [AACategory.Severe]: ' >118 KM/H',
  [AACategory.Moderate]: ' >89 KM/H',
};
