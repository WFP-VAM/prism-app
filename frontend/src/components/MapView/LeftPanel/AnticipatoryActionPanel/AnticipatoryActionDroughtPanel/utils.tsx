/* eslint-disable react-refresh/only-export-components */
import {
  Checkbox,
  CheckboxProps,
  FormControlLabel,
  FormControlLabelProps,
  Radio,
  RadioProps,
  useRadioGroup,
  withStyles,
} from '@material-ui/core';
import { black, borderGray, lightGrey } from 'muiTheme';
import { useSafeTranslation } from 'i18n';
import {
  AACategoryType,
  AAPhaseType,
  AAcategory,
  AnticipatoryActionDataRow,
} from 'context/anticipatoryAction/AADroughtStateSlice/types';
import {
  LIGHT_BLUE_HEX,
  LIGHT_GREEN_HEX,
} from 'components/MapView/DateSelector/TimelineItems/utils';
import AAIcon from '../AAIcon';

const StyledRadio = withStyles({
  root: {
    '&$checked': {
      color: black,
    },
    padding: '0.25rem',
  },
})((props: RadioProps) => <Radio color="default" {...props} />);

export const StyledRadioLabel = withStyles({
  root: {
    border: `1px solid ${borderGray}`,
    borderRadius: '32px',
    height: '1.75rem',
    marginLeft: 0,
    marginRight: '0.5rem',
  },
})(({ label, ...props }: Omit<FormControlLabelProps, 'control'>) => {
  const { t } = useSafeTranslation();
  const radioGroup = useRadioGroup();
  const checked = radioGroup?.value === props.value;

  const colorTags: { [key: string]: string } = {
    'Window 1': LIGHT_BLUE_HEX,
    'Window 2': LIGHT_GREEN_HEX,
  };

  const color = colorTags[label as string] || undefined;

  return (
    <FormControlLabel
      style={{ background: checked ? lightGrey : undefined }}
      label={
        <span style={{ marginRight: '1rem' }}>
          {color ? (
            <span
              style={{
                display: 'inline-block',
                width: '10px',
                height: '10px',
                backgroundColor: color,
                marginRight: '0.5rem',
              }}
            />
          ) : null}
          {typeof label === 'string' ? t(label) : label}
        </span>
      }
      control={<StyledRadio />}
      {...props}
    />
  );
});

const StyledCheckbox = withStyles({
  root: {
    '&$checked': {
      color: black,
    },
    padding: '0.2rem',
  },
})((props: CheckboxProps) => <Checkbox color="default" {...props} />);

export const StyledCheckboxLabel = withStyles({
  root: {
    border: `1px solid ${borderGray}`,
    borderRadius: '2px',
    height: '1.75rem',
    marginLeft: 0,
  },
})(
  ({
    label,
    checkBoxProps,
    ...props
  }: Omit<FormControlLabelProps, 'control'> & {
    checkBoxProps: CheckboxProps;
  }) => (
    <FormControlLabel
      style={{ background: checkBoxProps.checked ? lightGrey : undefined }}
      label={<span style={{ marginRight: '0.5rem' }}>{label}</span>}
      control={<StyledCheckbox {...checkBoxProps} />}
      {...props}
    />
  ),
);

const AACategoryPhaseMap: { [key: string]: any } = {
  na: {
    color: '#F1F1F1',
    iconProps: { topText: 'na', bottomText: '-', color: 'black' },
  },
  ny: {
    color: '#F1F1F1', // Note: Special handling required for 'ny' in getAAColor for non-layer usage
    iconProps: { topText: 'ny', bottomText: '-', color: 'black' },
  },
  Severe: {
    Set: {
      color: '#831F00',
      iconProps: { topText: 'S', bottomText: 'SEV', color: 'white' },
    },
    Ready: {
      color: '#E63701',
      iconProps: { topText: 'R', bottomText: 'SEV', color: 'white' },
    },
    na: {
      color: '#F1F1F1',
      iconProps: { topText: 'na', bottomText: 'SEV', color: 'black' },
    },
  },
  Moderate: {
    Set: {
      color: '#FF8934',
      iconProps: { topText: 'S', bottomText: 'MOD', color: 'black' },
    },
    Ready: {
      color: '#FFD52D',
      iconProps: { topText: 'R', bottomText: 'MOD', color: 'black' },
    },
    na: {
      color: '#F1F1F1',
      iconProps: { topText: 'na', bottomText: 'MOD', color: 'black' },
    },
  },
  Mild: {
    Set: {
      color: '#FFF503',
      iconProps: { topText: 'S', bottomText: 'MIL', color: 'black' },
    },
    Ready: {
      color: '#FFFCB3',
      iconProps: { topText: 'R', bottomText: 'MIL', color: 'black' },
    },
    na: {
      color: '#F1F1F1',
      iconProps: { topText: 'na', bottomText: 'MIL', color: 'black' },
    },
  },
  Normal: {
    Set: {
      color: '#FFF503',
      iconProps: { topText: 'S', bottomText: 'BNO', color: 'black' },
    },
    Ready: {
      color: '#FFFCB3',
      iconProps: { topText: 'R', bottomText: 'BNO', color: 'black' },
    },
    na: {
      color: '#F1F1F1',
      iconProps: { topText: 'na', bottomText: 'BNO', color: 'black' },
    },
  },
};

export function getAAColor(
  category: AACategoryType,
  phase: AAPhaseType,
  forLayer: boolean = false,
) {
  const categoryData = AACategoryPhaseMap[category];
  if (category === 'ny' && !forLayer) {
    return `repeating-linear-gradient(
      -45deg,
      #F1F1F1,
      #F1F1F1 10px,
      white 10px,
      white 20px
    )`;
  }
  if (categoryData.color) {
    return categoryData.color;
  }
  const phaseData = categoryData[phase];
  if (!phaseData) {
    throw new Error(`Icon not implemented: ${category}, ${phase}`);
  }
  return phaseData.color;
}

export function getAAIcon(
  category: AACategoryType,
  phase: AAPhaseType,
  forLayer?: boolean,
) {
  const background = getAAColor(category, phase, forLayer);

  const categoryData = AACategoryPhaseMap[category];
  if (categoryData.iconProps) {
    const iconProps = forLayer
      ? { ...categoryData.iconProps, bottomText: undefined }
      : categoryData.iconProps;
    return (
      <AAIcon
        background={background}
        fillBackground={!forLayer}
        {...iconProps}
      />
    );
  }
  const phaseData = categoryData[phase];
  if (!phaseData) {
    throw new Error(`Icon not implemented: ${category}, ${phase}`);
  }
  return (
    <AAIcon
      background={background}
      fillBackground={!forLayer}
      {...phaseData.iconProps}
    />
  );
}

export function AADataSeverityOrder(
  category: AnticipatoryActionDataRow['category'],
  phase: AnticipatoryActionDataRow['phase'],
  bonus: number = 100,
) {
  const catIndex = AAcategory.findIndex(x => x === category);
  const phaseBonus = phase === 'Set' ? bonus : 0;

  return catIndex * 10 + phaseBonus;
}
