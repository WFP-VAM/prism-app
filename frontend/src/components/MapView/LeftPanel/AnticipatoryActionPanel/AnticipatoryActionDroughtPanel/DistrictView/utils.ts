import {
  AnticipatoryActionDataRow,
  AnticipatoryActionState,
} from 'context/anticipatoryAction/AADroughtStateSlice/types';
import { getSeason } from 'context/anticipatoryAction/AADroughtStateSlice/utils';
import { AADataSeverityOrder } from '../utils';
import { TimelineRow } from '../Timeline/utils';

export function districtViewTransform(
  data: AnticipatoryActionDataRow[] | undefined,
  filters: AnticipatoryActionState['filters'],
) {
  if (!data) {
    return undefined;
  }
  const { categories: categoryFilters, selectedIndex, selectedDate } = filters;
  const season = getSeason(selectedDate);

  // Keep valid data only and switch SET phase to 'na' if the READY was not valid
  const validData = data
    .filter(
      x =>
        categoryFilters[x.category] &&
        season === x.season &&
        (selectedIndex === '' || x.index === selectedIndex),
    )
    .map(x => {
      if (x.isValid) {
        return x;
      }
      return {
        ...x,
        phase: 'na',
      } as AnticipatoryActionDataRow;
    });

  const monthsMap = new Map<string, AnticipatoryActionDataRow[]>();
  validData.forEach(x => {
    const val = monthsMap.get(x.date);
    monthsMap.set(x.date, val ? [...val, x] : [x]);
  });

  // eslint-disable-next-line fp/no-mutating-methods
  const months = Array.from(monthsMap.keys()).sort();

  const topFiltered = months
    .map(date => {
      const dateData = monthsMap.get(date);
      if (!dateData) {
        // this should never happen
        throw new Error('Invalid Date');
      }

      if (dateData.every(x => x.computedRow)) {
        return [];
      }

      const setCategories = new Set(
        dateData.filter(x => x.phase === 'Set').map(x => x.category),
      );

      const readyCategories = new Set(
        dateData.filter(x => x.phase === 'Ready').map(x => x.category),
      );

      // only keep data for the highest phase by category
      const ret = dateData.filter(
        x =>
          x.phase === 'Set' ||
          (x.phase === 'Ready' && !setCategories.has(x.category)) ||
          (x.phase === 'na' &&
            !setCategories.has(x.category) &&
            !readyCategories.has(x.category)),
      );

      return ret;
    })
    .flat();

  const sevMap = new Map<number, AnticipatoryActionDataRow[]>();
  topFiltered.forEach(x => {
    // return category data on one line by setting phase bonus to 0
    const sevVal = AADataSeverityOrder(x.category, x.phase, 0);
    const val = sevMap.get(sevVal);
    sevMap.set(sevVal, val ? [...val, x] : [x]);
  });

  const transformed = Object.fromEntries(sevMap);

  return {
    months: Object.fromEntries(
      [...new Set(topFiltered.map(x => x.date))].map(x => [
        x,
        new Date(x).toLocaleString('en-US', { month: 'short' }),
      ]),
    ),
    transformed,
  };
}

export function dateSorter(
  a: [string, AnticipatoryActionDataRow[] | TimelineRow],
  b: [string, AnticipatoryActionDataRow[] | TimelineRow],
) {
  if (a[0] > b[0]) {
    return -1;
  }
  if (a[0] < b[0]) {
    return 1;
  }
  return 0;
}
