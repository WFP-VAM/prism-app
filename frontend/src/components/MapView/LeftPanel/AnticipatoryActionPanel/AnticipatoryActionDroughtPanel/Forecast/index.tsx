import {
  Button,
  Typography,
  createStyles,
  makeStyles,
} from '@material-ui/core';
import ChartDataLabels from 'chartjs-plugin-datalabels';
import { Scatter } from 'react-chartjs-2';
import { useDispatch, useSelector } from 'react-redux';
import { lightGrey } from 'muiTheme';
import {
  AADataSelector,
  AAFiltersSelector,
  AASelectedDistrictSelector,
  setAAView,
} from 'context/anticipatoryAction/AADroughtStateSlice';
import {
  AACategoryType,
  AAView,
  AAcategory,
} from 'context/anticipatoryAction/AADroughtStateSlice/types';
import { dateRangeSelector } from 'context/mapStateSlice/selectors';
import { useSafeTranslation } from 'i18n';
import { ClearAll, Reply } from '@material-ui/icons';
import { getFormattedDate } from 'utils/date-utils';
import { getAAColor } from '../utils';
import { useAACommonStyles } from '../../utils';
import { chartOptions, forecastTransform, getChartData } from './utils';

interface ForecastProps {
  dialogs: {
    text: string;
    onclick: () => void;
  }[];
}

function Forecast({ dialogs }: ForecastProps) {
  const classes = useForecastStyle();
  const commonClasses = useAACommonStyles();
  const dispatch = useDispatch();
  const { t } = useSafeTranslation();
  const AAData = useSelector(AADataSelector);
  const selectedDistrict = useSelector(AASelectedDistrictSelector);
  const filters = useSelector(AAFiltersSelector);
  const { startDate: selectedDate } = useSelector(dateRangeSelector);

  const { chartData, indexes } = forecastTransform({
    data: AAData,
    filters,
    selectedDistrict,
  });

  const forecastButtons = [
    {
      icon: Reply,
      text: 'Back',
      onClick: () => dispatch(setAAView(AAView.District)),
    },
    {
      icon: ClearAll,
      text: t('Timeline'),
      onClick: () => dispatch(setAAView(AAView.Timeline)),
    },
  ];

  if (indexes.length === 0) {
    return (
      <div className={classes.noData}>
        <Typography>
          {t(
            `No data available yet for ${selectedDistrict}. Please pick a later date.`,
          )}
        </Typography>
      </div>
    );
  }

  return (
    <>
      <Typography variant="h3" style={{ marginLeft: '1rem' }}>
        {t('Forecast data as of ')}
        {getFormattedDate(selectedDate, 'locale', t('date_locale'))}
      </Typography>
      <div className={classes.charts}>
        <div className={classes.chartsHeader}>
          <div style={{ minWidth: '3rem' }} />
          {indexes.map(x => (
            <Typography key={x} className={classes.label}>
              {t(x)}
            </Typography>
          ))}
          <div style={{ minWidth: '10px' }} />
        </div>

        {
          // eslint-disable-next-line fp/no-mutating-methods
          Object.entries(chartData)
            .filter(([sev, _]) => filters.categories[sev as AACategoryType])
            .sort((a, b) => {
              const indexA = AAcategory.findIndex(x => x === a[0]);
              const indexB = AAcategory.findIndex(x => x === b[0]);

              if (indexA > indexB) {
                return -1;
              }
              if (indexA < indexB) {
                return 1;
              }
              return 0;
            })
            .map(([sev, indexData]) => {
              const color = getAAColor(sev as AACategoryType, 'Ready');
              return (
                <div className={classes.chartLine} key={sev}>
                  <div
                    className={classes.textWrap}
                    style={{ backgroundColor: color }}
                  >
                    <Typography
                      className={classes.text}
                      style={{
                        color: sev === 'Severe' ? 'white' : ' black',
                        border: `1px solid ${
                          sev === 'Severe' ? 'white' : ' black'
                        }`,
                      }}
                    >
                      {t(sev)}
                    </Typography>
                  </div>

                  <div className={classes.chartWrapper}>
                    <Scatter
                      data={getChartData(indexData, color) as any}
                      plugins={[ChartDataLabels]}
                      options={chartOptions as any}
                    />
                  </div>
                </div>
              );
            })
        }
      </div>
      <div className={commonClasses.footerWrapperVert}>
        <div className={commonClasses.footerActionsWrapper}>
          {forecastButtons.map(x => (
            <Button
              key={x.text}
              className={commonClasses.footerButton}
              variant="outlined"
              fullWidth
              onClick={x.onClick}
              startIcon={<x.icon />}
            >
              <Typography>{t(x.text)}</Typography>
            </Button>
          ))}
        </div>
        <div className={commonClasses.footerDialogsWrapperVert}>
          {dialogs.map(dialog => (
            <Typography
              key={dialog.text}
              className={commonClasses.footerDialog}
              component="button"
              onClick={() => dialog.onclick()}
            >
              {t(dialog.text)}
            </Typography>
          ))}
        </div>
      </div>
    </>
  );
}

const useForecastStyle = makeStyles(() =>
  createStyles({
    noData: {
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      height: '100%',
    },
    charts: {
      display: 'flex',
      flexDirection: 'column',
      gap: '1rem',
      width: '100%',
      background: lightGrey,
    },
    chartsHeader: {
      display: 'flex',
      flexDirection: 'row',
      flexWrap: 'nowrap',
      paddingLeft: '2.6rem',
      paddingRight: '0.5rem',
      marginBottom: '-1rem',
      background: 'white',
    },
    chartLine: {
      display: 'flex',
      flexDirection: 'row',
      flexWrap: 'nowrap',
      background: 'white',
    },
    textWrap: {
      width: '2.6rem',
      display: 'flex',
      justifyContent: 'center',
    },
    text: {
      fontSize: '0.9rem',
      fontWeight: 400,
      borderRadius: '2px',
      writingMode: 'vertical-lr',
      textTransform: 'uppercase',
      transform: 'rotate(180deg)',
      padding: '0.5rem 0.1rem',
      margin: 'auto',
    },
    chartWrapper: {
      paddingBottom: '0.5rem',
      height: '7rem',
      width: '100%',
    },
    label: {
      background: lightGrey,
      margin: '0.5rem',
      borderRadius: '4px',
      textAlign: 'center',
      textTransform: 'uppercase',
      lineHeight: '2rem',
      width: '100%',
    },
  }),
);

export default Forecast;
