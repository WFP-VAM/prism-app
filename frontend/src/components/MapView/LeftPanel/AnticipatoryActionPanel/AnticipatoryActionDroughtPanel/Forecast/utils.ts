import {
  AACategoryType,
  AnticipatoryActionDataRow,
  AnticipatoryActionState,
} from 'context/anticipatoryAction/AADroughtStateSlice/types';
import { getSeason } from 'context/anticipatoryAction/AADroughtStateSlice/utils';

const indexOrder = ['SPI', 'DRY'];
const monthsOrder = ['O', 'N', 'D', 'J', 'F', 'M', 'A', 'M', 'J'];

function sortIndexes(indexes: string[]): string[] {
  // eslint-disable-next-line fp/no-mutating-methods
  return indexes.sort((a, b) => {
    const [typeA, monthA] = a.split(' ');
    const [typeB, monthB] = b.split(' ');

    const typeIndexA = indexOrder.indexOf(typeA);
    const typeIndexB = indexOrder.indexOf(typeB);

    if (typeIndexA !== typeIndexB) {
      return typeIndexA - typeIndexB;
    }

    const monthIndexA = monthsOrder.findIndex(month =>
      monthA.startsWith(month),
    );
    const monthIndexB = monthsOrder.findIndex(month =>
      monthB.startsWith(month),
    );

    return monthIndexA - monthIndexB;
  });
}

interface ForecastTransformParams {
  filters: AnticipatoryActionState['filters'];
  selectedDistrict: AnticipatoryActionState['selectedDistrict'];
  data: AnticipatoryActionState['data'];
}

interface IndexData {
  probability: number | null;
  showWarningSign: boolean;
}

interface CategoryData {
  [index: string]: IndexData;
}

interface ChartData {
  [category: string]: CategoryData;
}

export function forecastTransform({
  filters,
  selectedDistrict,
  data,
}: ForecastTransformParams) {
  const { selectedWindow, selectedDate } = filters;
  const season = getSeason(selectedDate);

  const dateData = (
    selectedWindow === 'All'
      ? [
          ...(data['Window 1'][selectedDistrict] || []),
          ...(data['Window 2'][selectedDistrict] || []),
        ]
      : data[selectedWindow][selectedDistrict] || []
  ).filter(
    x => !selectedDate || (x.date <= selectedDate && x.season === season),
  );

  // eslint-disable-next-line fp/no-mutating-methods
  const indexes = sortIndexes([...new Set(dateData.map(x => x.index))]);

  const sevMap = new Map<AACategoryType, AnticipatoryActionDataRow[]>();
  dateData.forEach(x => {
    const val = sevMap.get(x.category);
    sevMap.set(x.category, val ? [...val, x] : [x]);
  });
  const groupedBySev = Object.fromEntries(sevMap);

  const chartData: ChartData = Object.fromEntries(
    Object.entries(groupedBySev).map(([cat, catData]) => {
      const val = indexes.map(index => {
        const indexData = catData.filter(
          x => x.index === index && !x.computedRow,
        );
        if (indexData.length > 0) {
          // Sort by date in descending order to get the latest date first
          // eslint-disable-next-line fp/no-mutating-methods
          indexData.sort((a, b) => b.date.localeCompare(a.date));
          // Take the probability of the first element (latest date)
          const latest = parseFloat(
            ((indexData[0].probability || 0) * 100).toFixed(2),
          );

          const showWarningSign = Boolean(indexData[0].isValid);
          return [index, { probability: latest, showWarningSign }];
        }
        return [index, { probability: null, showWarningSign: false }];
      });
      return [cat, Object.fromEntries(val)];
    }),
  );

  return { chartData, indexes };
}

export const chartOptions = {
  legend: {
    display: false,
  },
  tooltips: {
    enabled: false,
  },
  layout: {
    padding: {
      left: 10,
      right: 10,
      top: 15,
    },
  },
  maintainAspectRatio: false,
  scales: {
    xAxes: [
      {
        gridLines: {
          display: false,
        },
        ticks: {
          suggestedMin: 0,
          display: false,
          stepSize: 1,
        },
      },
    ],
    yAxes: [
      {
        ticks: {
          suggestedMin: 0,
          suggestedMax: 40,
          stepSize: 10,
          callback: (value: number, _index: number, _values: number[]) =>
            `${value}%`,
        },
      },
    ],
  },
  plugins: {
    datalabels: {
      align: 'left',
      anchor: 'right',
      color: (context: any) => context.dataset.backgroundColor,
    },
    legend: {
      labels: {
        usePointStyle: true,
      },
    },
  },
};

export const getChartData = (
  indexes: {
    [key: string]: {
      probability?: number | null;
      showWarningSign: boolean | null;
    } | null;
  },
  backgroundColor: string,
) => ({
  labels: Object.keys(indexes),
  datasets: [
    {
      data: Object.entries(indexes).map(([_index, val], i) => ({
        x: i + 0.6,
        y: val?.probability,
        z: val?.showWarningSign,
      })),
      // Triangle pointer
      pointStyle: 'triangle',
      rotation: 90,
      radius: 10,
      hoverRadius: 10,
      backgroundColor,
      // Label
      datalabels: {
        labels: {
          value: {
            offset: -2, // offset from the point
            align: 'left',
            backgroundColor: 'white',
            borderColor: (ctx: any) => ctx.dataset.backgroundColor,
            borderWidth: 1,
            borderRadius: 2,
            color: 'black',
            formatter: (value: any, _ctx: any) =>
              `${value.z ? '⚠️ ' : ''}${value.y}%`,
          },
        },
      },
    },
  ],
});
