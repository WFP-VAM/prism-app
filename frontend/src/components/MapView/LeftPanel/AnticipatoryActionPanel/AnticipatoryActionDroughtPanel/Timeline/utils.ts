import { AAWindowKeys } from 'config/utils';
import {
  AACategoryType,
  AAPhase,
  AAPhaseType,
  AAcategory,
  AnticipatoryActionDataRow,
  AnticipatoryActionState,
} from 'context/anticipatoryAction/AADroughtStateSlice/types';
import { getSeason } from 'context/anticipatoryAction/AADroughtStateSlice/utils';

function getColumnKey(val: AnticipatoryActionDataRow): number {
  const { category, phase } = val;
  const catIndex = AAcategory.findIndex(x => x === category);
  const phaseIndex = AAPhase.findIndex(x => x === phase);
  return catIndex * 10 + phaseIndex;
}

export type TimelineRow = {
  status: {
    category: AACategoryType;
    phase: AAPhaseType;
  };
  data: AnticipatoryActionDataRow[];
};

interface TimelineTransformParams {
  filters: AnticipatoryActionState['filters'];
  selectedDistrict: AnticipatoryActionState['selectedDistrict'];
  data: AnticipatoryActionState['data'];
}

export function timelineTransform({
  filters,
  selectedDistrict,
  data,
}: TimelineTransformParams) {
  const { selectedWindow, selectedIndex, categories, selectedDate } = filters;
  const season = getSeason(selectedDate);

  const windowData = (
    selectedWindow === 'All' ? AAWindowKeys : [selectedWindow]
  ).map(win => {
    const districtData = !!selectedDistrict && data[win][selectedDistrict];
    if (!districtData) {
      return [win, null];
    }

    const filtered = districtData.filter(
      x =>
        x.season === season &&
        !x.computedRow &&
        (selectedIndex === '' || selectedIndex === x.index) &&
        categories[x.category],
    );

    const months = [...new Set(filtered.map(x => x.date))].map(x => [
      x,
      new Date(x).toLocaleString('en-US', { month: 'short' }),
    ]);

    const categoriesMap = new Map<
      number,
      {
        status: { category: AACategoryType; phase: AAPhaseType };
        data: AnticipatoryActionDataRow[];
      }
    >();
    filtered.forEach(x => {
      const key = getColumnKey(x);
      const val = categoriesMap.get(key);
      if (val) {
        // Sort the new data based on validity
        // eslint-disable-next-line fp/no-mutating-methods
        const sortedData = [
          x,
          ...val.data.filter(item => item.date === x.date),
        ].sort((a, b) => {
          if (a.isValid && a.isOtherPhaseValid) {
            return -1;
          }
          if (b.isValid && b.isOtherPhaseValid) {
            return 1;
          }
          if (a.isValid) {
            return -1;
          }
          if (b.isValid) {
            return 1;
          }
          return 0;
        });

        // Keep only the highest priority item for this date
        const highestPriorityItem = sortedData[0];

        // Update the data array
        // eslint-disable-next-line fp/no-mutation
        val.data = [
          ...val.data.filter(item => item.date !== x.date),
          highestPriorityItem,
        ];
      } else {
        categoriesMap.set(key, {
          status: {
            category: x.category,
            phase: x.phase,
          },
          data: [x],
        });
      }
    });
    return [win, { months, rows: Object.fromEntries(categoriesMap) }];
  }) as [
    (typeof AAWindowKeys)[number],
    {
      months: string[][];
      rows: { [id: string]: TimelineRow };
    } | null,
  ][];

  const allRows = windowData
    .map(([_win, x]) =>
      Object.entries(x?.rows || {}).map(([id, info]) => [
        id,
        { status: info.status, data: [] },
      ]),
    )
    .flat() as [string, TimelineRow][];

  return {
    windowData: Object.fromEntries(windowData),
    allRows: Object.fromEntries(allRows),
  };
}
