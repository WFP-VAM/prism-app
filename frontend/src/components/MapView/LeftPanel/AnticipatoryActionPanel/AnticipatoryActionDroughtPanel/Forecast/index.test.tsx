import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { AnticipatoryActionState } from 'context/anticipatoryAction/AADroughtStateSlice/types';
import { Panel } from 'config/types';
import { defaultDialogs, mockAAData } from '../test.utils';
import { forecastTransform } from './utils';
import Forecast from '.';

const mockStore = configureStore([]);

const filters: AnticipatoryActionState['filters'] = {
  selectedWindow: 'Window 1',
  categories: {
    Severe: true,
    Moderate: true,
    Mild: true,
    na: true,
    ny: true,
    Normal: true,
  },
  selectedIndex: '',
  selectedDate: undefined,
};

const store = mockStore({
  mapState: {
    layers: [],
    dateRange: { startDate: 1701432000000 },
    maplibreMap: () => {},
    errors: [],
    layersData: [],
    loadingLayerIds: [],
    boundaryRelationData: {},
  },
  serverState: { availableDates: {}, loading: false },
  anticipatoryActionDroughtState: {
    filters,
    selectedDistrict: 'Changara',
    monitoredDistricts: ['Changara'],
    data: mockAAData,
  },
  leftPanelState: {
    tabValue: Panel.AnticipatoryActionDrought,
  },
});

test('renders as expected', () => {
  const { container } = render(
    <BrowserRouter>
      <Provider store={store}>
        <Forecast dialogs={defaultDialogs} />
      </Provider>
    </BrowserRouter>,
  );
  expect(container).toMatchSnapshot();
});

test('District view transformation', () => {
  const res = forecastTransform({
    data: mockAAData,
    filters,
    selectedDistrict: 'Changara',
  });

  expect(res).toEqual(out);
});

const out = {
  chartData: {
    Mild: {
      'SPI DJF': {
        probability: 28,
        showWarningSign: true,
      },
    },
    Moderate: {
      'SPI DJF': {
        probability: 22,
        showWarningSign: false,
      },
    },
  },
  indexes: ['SPI DJF'],
};
