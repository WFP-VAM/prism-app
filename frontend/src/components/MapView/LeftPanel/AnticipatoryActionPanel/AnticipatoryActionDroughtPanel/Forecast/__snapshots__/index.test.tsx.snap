// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <mock-typography
    style="margin-left: 1rem;"
    variant="h3"
  >
    Forecast data as of 
    December 1, 2023
  </mock-typography>
  <div
    class="makeStyles-charts-2"
  >
    <div
      class="makeStyles-chartsHeader-3"
    >
      <div
        style="min-width: 3rem;"
      />
      <mock-typography
        classname="makeStyles-label-8"
      >
        SPI DJF
      </mock-typography>
      <div
        style="min-width: 10px;"
      />
    </div>
    <div
      class="makeStyles-chartLine-4"
    >
      <div
        class="makeStyles-textWrap-5"
        style="background-color: rgb(255, 213, 45);"
      >
        <mock-typography
          classname="makeStyles-text-6"
          style="color: black; border: 1px solid black;"
        >
          Moderate
        </mock-typography>
      </div>
      <div
        class="makeStyles-chartWrapper-7"
      >
        <canvas
          height="150"
          width="300"
        />
      </div>
    </div>
    <div
      class="makeStyles-chartLine-4"
    >
      <div
        class="makeStyles-textWrap-5"
        style="background-color: rgb(255, 252, 179);"
      >
        <mock-typography
          classname="makeStyles-text-6"
          style="color: black; border: 1px solid black;"
        >
          Mild
        </mock-typography>
      </div>
      <div
        class="makeStyles-chartWrapper-7"
      >
        <canvas
          height="150"
          width="300"
        />
      </div>
    </div>
  </div>
  <div
    class="makeStyles-footerWrapperVert-14"
  >
    <div
      class="makeStyles-footerActionsWrapper-10"
    >
      <mock-button
        classname="makeStyles-footerButton-12"
        fullwidth="true"
        starticon="[object Object]"
        variant="outlined"
      >
        <mock-typography>
          Back
        </mock-typography>
      </mock-button>
      <mock-button
        classname="makeStyles-footerButton-12"
        fullwidth="true"
        starticon="[object Object]"
        variant="outlined"
      >
        <mock-typography>
          Timeline
        </mock-typography>
      </mock-button>
    </div>
    <div
      class="makeStyles-footerDialogsWrapperVert-15"
    >
      <mock-typography
        classname="makeStyles-footerDialog-13"
        component="button"
      >
        How to read this screen
      </mock-typography>
    </div>
  </div>
</div>
`;
